app/main.py

```diff
@@ -0,0 +1,474 @@
+                sys.exit("!!!Exiting. Please set REDIS_HOST in env & restart.")
+
+    def health(self):
+        """Return a ping response from redis client."""
+        return self.redis.ping()
+
+    def get(self, key):
+        """Get device from redit if it exists, otherwise load with DEFAULT."""
+        with self.redis.pipeline() as pipe:
+            while True:
+                try:
+                    pipe.watch(key)
+                    if pipe.exists(key) == 0:
+                        logging.warning("get: no device")
+                        pipe.multi()
+                        pipe.hmset(key, DEFAULT)
+                        pipe.execute()
+                        device = DEFAULT
+                        break
+                    else:
+                        device = pipe.hgetall(key)
+                        # Initialising empty dictionary
+                        new_dict = {}
+
+                        # Convert dictionary items from bytes to strings
+                        for dkey, value in device.items():
+                            new_dict[dkey.decode(
+                                "utf-8")] = value.decode("utf-8")
+                        device = new_dict
+                        logging.debug("get: device found: %s", str(device))
+                        break
+                except WatchError:
+                    logging.warning("get: watcherror, trying again")
+                    continue
+                except redis.ConnectionError:
+                    logging.exception("Connection error")
+                    return "Error: error connectiong to Redis host"
+
+                except Exception:
+                    logging.exception("Unexpected error")
+
+        return device
+
+    def set(self, key, field, value):
+        """Set value if device exists, else first initialize device to DEFAULT."""
+        with self.redis.pipeline() as pipe:
+            while True:
+                try:
+                    pipe.watch(key)
+                    if pipe.exists(key) == 0:
+                        logging.warning("set: no device")
+                        pipe.hmset(key, DEFAULT)
+                        pipe.hmset(key, {field: value})
+                        device = pipe.hgetall(key)
+                        pipe.execute()
+                        logging.info("set: device now: %s", str(device))
+                        break
+
+                    else:
+                        pipe.hmset(key, {field: value})
+                        device = pipe.hgetall(key)
+                        logging.info("set: device found: %s", str(device))
+                        break
+
+                except WatchError:
+                    logging.warning("set: watcherror, trying again")
+                    continue
+                except ConnectionError:
+                    logging.exception("Connection error")
+                except Exception:
+                    logging.exception("Unexpected error occured")
```

Comment:
Does this get stuck in an infinite loop if redis is down or something else causes an exception to be raised?

Reply:
This also should be address in new version in master.